ts/client/src/accounts/mangoAccount.ts

```diff
@@ -1628,6 +1676,243 @@ export class PerpPosition {
+        break;
+      }
+      prevLp = nextLp;
+    }
+    return prevLp;
+  }
+
+  public static getAggregatePerpOrderSizeAndQuoteNative(
+    perpMarket: PerpMarket,
+    orders: PerpOrder[],
+    side: PerpOrderSide,
+    tillPriceExclusive: I80F48,
+  ): { sizeLots: BN; quoteNative: I80F48 } {
+    const filtered = orders.filter((oo) =>
+      side == PerpOrderSide.bid
+        ? perpMarket.priceLotsToNative(oo.priceLots).gte(tillPriceExclusive)
+        : perpMarket.priceLotsToNative(oo.priceLots).lte(tillPriceExclusive),
+    );
+    return {
+      sizeLots: filtered.reduce((sum, a) => sum.add(a.sizeLots), new BN(0)),
+      quoteNative: filtered.reduce(
+        (sum, a) =>
+          sum.add(
+            I80F48.fromI64(a.sizeLots)
+              .mul(I80F48.fromI64(perpMarket.baseLotSize))
+              .mul(perpMarket.priceLotsToNative(a.priceLots)),
+          ),
+        ZERO_I80F48(),
+      ),
+    };
+  }
+
+  public getLiquidationPriceInternal(
+    group: Group,
+    perpMarket: PerpMarket,
+    mangoAccount: MangoAccount,
+    sortedOo: PerpOrder[],
+    tillPriceExclusive: I80F48,
+  ): I80F48 {
+    const settleTokenBank = group.getFirstBankByTokenIndex(
+      perpMarket.settleTokenIndex,
+    );
+    const settleTokenPrice = settleTokenBank.price;
+
+    // For constants B, T, C and final liqudation price equation, refer to
+    // https://www.notion.so/mango-markets/Mango-v4-Perp-Liq-Price-b7dde5e70c724d07934a21dc0aa79518?pvs=4
+
+    // Set base and quote
+    // Take into account bids or asks if they'd execute
+    // before we hit tillPriceExclusive i.e. previously computed liquidation price
+    let wbpn; // B
+    let qpn = this.quotePositionNative.sub(
+      this.getUnsettledFunding(perpMarket),
+    );
+    {
+      const baseNative = this.getBasePositionNative(perpMarket);
+      if (baseNative.isZero()) {
+        return MINUS_ONE_I80F48();
+      }
+
+      if (baseNative.isNeg()) {
+        const ret = PerpPosition.getAggregatePerpOrderSizeAndQuoteNative(
+          perpMarket,
+          sortedOo,
+          PerpOrderSide.ask,
+          tillPriceExclusive,
+        );
+        wbpn = perpMarket.maintBaseLiabWeight.mul(
+          baseNative.sub(
+            I80F48.fromI64(ret.sizeLots).mul(
+              I80F48.fromI64(perpMarket.baseLotSize),
+            ),
+          ),
+        );
+        qpn = qpn.add(ret.quoteNative);
+      } else {
+        const ret = PerpPosition.getAggregatePerpOrderSizeAndQuoteNative(
+          perpMarket,
+          sortedOo,
+          PerpOrderSide.ask,
+          tillPriceExclusive,
+        );
+        wbpn = perpMarket.maintBaseAssetWeight.mul(
+          baseNative.add(
+            I80F48.fromI64(ret.sizeLots).mul(
+              I80F48.fromI64(perpMarket.baseLotSize),
+            ),
+          ),
+        );
+        qpn = qpn.sub(ret.quoteNative);
+      }
+    }
+
+    if (wbpn.isZero()) {
+      return MINUS_ONE_I80F48();
```

Comment:
If it always returns -1, even for shorts, then shouldn't that mean that prevLp will be -1 and it won't consider what happens when the asks > oracle price execute?